packages/core/src/indexing/service.ts

```diff
@@ -1,1187 +1,724 @@
+      typeof source.criteria.address === "string"
+        ? source.criteria.address
+        : undefined;
+
+    if (contractsByChainId[source.chainId] === undefined) {
+      contractsByChainId[source.chainId] = {};
+    }
 
-  /**
-   * This method is triggered by the realtime sync service detecting a reorg,
-   * which can happen at any time. The event queue and the indexing store can be
-   * in one of several different states that we need to keep in mind:
-   *
-   * 1) No events have been added to the queue yet.
-   * 2) No unsafe events have been processed (checkpoint <= safeCheckpoint).
-   * 3) Unsafe events may have been processed (checkpoint > safeCheckpoint).
-   * 4) The queue has encountered a user error and is waiting for a reload.
-   *
-   * Note: It's crucial that we acquire all mutex locks while handling the reorg.
-   * This will only ever run while the queue is idle, so we can be confident
-   * that checkpoint matches the current state of the indexing store,
-   * and that no unsafe events will get processed after handling the reorg.
-   *
-   * Note: Caller should (probably) immediately call processEvents after this method.
-   */
-  handleReorg = async (safeCheckpoint: Checkpoint) => {
-    await this.loadingMutex.runExclusive(async () => {
-      try {
-        const hasProcessedInvalidEvents = Object.values(
-          this.indexingFunctionStates,
-        ).some((state) =>
-          isCheckpointGreaterThan(
-            state.tasksProcessedToCheckpoint,
-            safeCheckpoint,
-          ),
-        );
-
-        if (!hasProcessedInvalidEvents) {
-          // No unsafe events have been processed, so no need to revert (case 1 & case 2).
-          this.common.logger.debug({
-            service: "indexing",
-            msg: "No unsafe events were detected while reconciling a reorg, no-op",
-          });
-          return;
-        }
-
-        // Unsafe events have been processed, must revert the indexing store and update
-        // eventsProcessedToTimestamp accordingly (case 3).
-        await this.indexingStore.revert({ checkpoint: safeCheckpoint });
-
-        this.common.metrics.ponder_indexing_completed_timestamp.set(
-          safeCheckpoint.blockTimestamp,
-        );
-
-        // Note: There's currently no way to know how many events are "thrown out"
-        // during the reorg reconciliation, so the event count metrics
-        // (e.g. ponder_indexing_processed_events) will be slightly inflated.
+    contractsByChainId[source.chainId][source.contractName] = {
+      abi: source.abi,
+      address: address ? checksumAddress(address) : address,
+      startBlock: source.startBlock,
+      endBlock: source.endBlock,
+      maxBlockRange: source.maxBlockRange,
+    };
+  }
 
-        this.common.logger.debug({
-          service: "indexing",
-          msg: `Reverted indexing store to safe timestamp ${safeCheckpoint.blockTimestamp}`,
+  // build db
+  const db = Object.keys(schema.tables).reduce<
```

Comment:
Up to you, but this file might be more readable if we moved this and the actions code back into their own files. You could just pass the `contextState` object as an argument.